mmv1/products/networkservices/WasmPlugin.yaml

```diff
@@ -0,0 +1,189 @@
+# Copyright 2024 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+---
+name: 'WasmPlugin'
+description: |
+  WasmPlugin is a resource representing a service executing a customer-provided Wasm module.
+min_version: 'beta'
+references:
+  guides:
+    'Configure a route extension': 'https://cloud.google.com/service-extensions/docs/create-plugin'
+  api: 'https://cloud.google.com/service-extensions/docs/reference/rest/v1beta1/projects.locations.wasmPlugins'
+docs:
+base_url: 'projects/{{project}}/locations/{{location}}/wasmPlugins'
+self_link: 'projects/{{project}}/locations/{{location}}/wasmPlugins/{{name}}'
+create_url: 'projects/{{project}}/locations/{{location}}/wasmPlugins?wasmPluginId={{name}}'
+create_verb: 'POST'
+update_verb: 'PATCH'
+update_mask: true
+read_query_params: '?view=WASM_PLUGIN_VIEW_FULL'
+timeouts:
+  insert_minutes: 20
+  update_minutes: 20
+  delete_minutes: 20
+async:
+  actions: ['create', 'delete', 'update']
+  type: 'OpAsync'
+  operation:
+    base_url: '{{op_id}}'
+  result:
+    resource_inside_response: false
+custom_code:
+examples:
+  - name: 'wasm_plugin_basic'
+    primary_resource_id: 'wasm_plugin'
+    min_version: 'beta'
+    vars:
+      wasm_plugin_name: 'my-wasm-plugin'
+parameters:
+  - name: 'location'
+    type: String
+    description: |
+      The location of the traffic extension
+    url_param_only: true
+    required: true
+    immutable: true
+  - name: 'name'
+    type: String
+    description: |
+      Identifier. Name of the WasmPlugin resource.
+    url_param_only: true
+    required: true
+    immutable: true
+properties:
+  - name: 'createTime'
+    type: Time
+    description: 'Output only. The timestamp when the resource was created.'
+    output: true
+  - name: 'updateTime'
+    type: Time
+    description: 'Output only. The timestamp when the resource was updated.'
+    output: true
+  - name: 'description'
+    type: String
+    description: |
+      Optional. A human-readable description of the resource.
+  - name: 'labels'
+    type: KeyValueLabels
+    description: 'Optional. Set of labels associated with the WasmPlugin resource.'
+  - name: 'mainVersionId'
+    type: ResourceRef
+    resource: 'WasmPluginVersion'
+    description: |
+      Optional. The ID of the WasmPluginVersion resource that is the currently serving one. The version referred to must be a child of this WasmPlugin resource and should be listed in the "versions" field.
```

Comment:
Is there a `google_network_services_wasm_plugin_version` resource on its way?

Reply:
I am currently trying to implement a version resource but I am having some issues since `main_version_id` appears to be required for creation (despite the [API docs](https://cloud.google.com/service-extensions/docs/reference/rest/v1alpha1/projects.locations.wasmPlugins#:~:text=count%22%3A%20%223%22%20%7D.-,mainVersionId,-string) describing it as optional) and needs to refer to a version listed in this plugin resource during request. Trying to list a initial version in the plugin resource and config additional versions as resources inevitably causes permadiffs.
  
Either way, I think this field would be better off as a String since it may not necessarily be used with a future versions resource.

Reply:
I just checked the API proto definition, and `main_version_id` is annotated with both `(google.api.field_behavior) = OPTIONAL` and a `validator.rule` that makes it required (and the latter wins).

The main reason I asked about `google_network_services_wasm_plugin_version` is if that resource can manage versions and so can a writeable `versions` field in `google_network_services_wasm_plugin` then there's no single canonical place to configure versions, which doesn't play well with Terraform.

Thinking about it a little more, I'm not sure that `google_network_services_wasm_plugin_version` would be possible, at least not without a lot of custom code or a sketchy user experience. Since the parent `google_network_services_wasm_plugin` needs to exist before creating a `google_network_services_wasm_plugin_version` child, the configuration sequence would be:

1. Configure a `google_network_services_wasm_plugin` with an empty `main_version_id` (assuming that the `validator.rule` is incorrect and can be removed from the API)
2. Create a `google_network_services_wasm_plugin_version`
3. Update the plugin from step 1 to set `main_version_id` to the version from step 2.

As you've discovered, step 3 is what makes things a PITA. There's a couple of ways to work around the problem:

- The easiest is not to implement `google_network_services_wasm_plugin_version` and make the `versions` field of `google_network_services_wasm_plugin` required instead.
- If `main_version_id` really is optional, remove the `validator.rule` (and wait for the subsequent API rollout). Then, implement `google_network_services_wasm_plugin_version` with an additional virtual boolean field, something like `main_version`. When a `google_network_services_wasm_plugin_version` is configured with `main_version = true` a custom post-create or post-update step would `PATCH` the parent plugin resource to set `main_version` to the version in question.

With the latter, `main_version_id` and `versions` in `google_network_services_wasm_plugin` can be output only or omitted entirely. You'll need custom encoder for `main_version_id` and `versions` for updates so that they don't get zeroed out when updating the other fields in the resource.

I think the separate version resource option is a more idiomatically Terraform approach and a better user experience, but it will be a fair amount of additional work to create and maintain. I'll leave it up to you which you prefer.

Reply:
`main_version_id` being required seems to be the intended behavior for the API, with versions being managed together with the WasmPlugin), so I will be keeping the versions implementation on the google_network_services_wasm_plugin resource.